### PR TITLE
Add relations to the cache, and extra model details.

### PR DESCRIPTION
**Pull request by howbazaar**
In order to provide the information to the upcoming model summary watcher (which will be implemented in the cache), we need to get some extra information into the cache.

There is a drive by that renames the ModelUpdate type from the core/multiwatcher package back to ModelInfo. This was renamed previously when the type was moved into the apiserver/params package and we had to deal with a naming collision. Renamed it back for this type for consistency with the rest of the package.

Adds three pieces of information from the state.Model in through the allwatcher: Cloud, Region, and Credential. Each of these is just a string value in the underlying document, so just adding these to the multiwatcher info type is sufficient to pass them along. Also needed to add these to the cache change type, and this makes them available to the internals of the cache.

The cache needed to know about relations, so pass the relation changes from the modelcache worker though to the controller, and add methods to add and remove the relations to the cached model.

## QA steps

Nothing is exposed to the outside world yet, that is coming.

## Documentation changes

Everything is internal, no user facing change.

### CODE REVIEW

File: core/cache/cachetest/state.go [hunk 1]
```diff
@@ -32,13 +35,22 @@ func ModelChange(c *gc.C, model *state.Model) cache.ModelChange {
 	status, err := model.Status()
 	c.Assert(err, jc.ErrorIsNil)
 
+	users, err := model.Users()
+	permissions := make(map[string]permission.Access)
+	for _, user := range users {
+		// Cache permission map is always lower case.
+		permissions[strings.ToLower(user.UserName)] = user.Access
```
Comment (wallyworld) [inline core/cache/cachetest/state.go]: This works for external users also? ie fred@external.
I can't recall if UserName includes the domain suffix

Comment (howbazaar) [reply core/cache/cachetest/state.go]: It does. UserName is the fred@external bit.

**Comment by howbazaar** [comment]
$$merge$$

**Comment by howbazaar** [comment]
$$merge$$ intermittent test failure

**Comment by howbazaar** [comment]
$$merge$$ different intermittent failure

**Pull request merged**